### PR TITLE
CI: Auto upload macOS/Linux releases to github

### PR DESCRIPTION
**Pull request by cg2121**
### Description
This automatically uploads releases when a new tag is published.

### Motivation and Context
The release process needs to automated.

### How Has This Been Tested?
Tested with CI on fork.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by WizardCM** [comment]
As with nightlies, these'd be unsigned (including the game capture DLLs). As we know, the macOS signing process is a lot of waiting and a bit of user input (purposefully not designed to be automated by Apple). I don't know the Windows signing well enough, but I assume that's best done manually too (after building?).

It's certainly possible to package the Windows build into an installer automatically, but that also depends on Jim's custom folder structure pre-package.

I personally think the PPA would be the best place to start automating on the CI, especially nightlies. That process is (from what I understand) pretty lengthy.

**Comment by PatTheMav** [comment]
Signing for macOS can be fully automated if we feel comfortable uploading our macOS developer keys to CI as a secret (AES-encrypted BASE64-string that is decrypted with a key stored as another secret).

I've implemented this once for obs-websocket: https://github.com/PatTheMav/obs-websocket/blob/a479f529af1e5bf4214d39cec36fad6e6f101a14/.github/workflows/tag_release.yml#L395

Also feel free to adopt the fully automated release step from obs-deps: https://github.com/PatTheMav/obs-deps/blob/427da4ca215d84aac5f1f0e28d700f1b581e564a/.github/workflows/build_deps.yml#L587

**Comment by RytoEX** [comment]
> Signing for macOS can be fully automated if we feel comfortable uploading our macOS developer keys to CI as a secret (AES-encrypted BASE64-string that is decrypted with a key stored as another secret).

We used to do this. See the old macOS build/package/deploy scripts.

**Comment by cg2121** [comment]
I've finally got it working properly.

**Comment by PatTheMav** [comment]
@cg2121 With #3444 and #3449 merged we have release-ready macOS builds created by CI now.

**Comment by cg2121** [comment]
I've currently simplified this to only upload Mac releases, as it is the only OS with the full build process in the Git workflow.

Windows and Linux can be done at a different time when their builds are more automated.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -655,3 +655,60 @@ jobs:
...
+      - name: 'Create Release'
+        id: create_release
+        uses: actions/create-release@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          tag_name: ${{ env.OBS_GIT_TAG }}
+          release_name: OBS Studio ${{ env.OBS_GIT_TAG }}
+          draft: false
+          prerelease: ${{ env.IS_RC }}
```
Comment (PatTheMav) [inline .github/workflows/main.yml:700]: I haven't checked the specifics, but as "IS_RC" is not used anywhere else, I'd replace this with a built-in expression instead. Saves the overhead of running those bash jobs.

Comment (cg2121) [reply .github/workflows/main.yml:729]: Removed the bash jobs.

**Comment by cg2121** [comment]
It is updated now to create the release in draft mode.

**Comment by jp9000** [comment]
There are some issues with this. First issue is that the file names are not going to be the desired filenames. For example, the 26.1.2 macOS release name was `obs-mac-26.1.2.dmg`. The filename that's generated by github on the other hand is mangled and has the commit hash in it. For example, 27 RC2 was `2021-04-12-dca1d3714-27.0.0-rc2-rel-macOS.dmg`. This is generally not ideal as a release filename.

Another issue dodgepong pointed out is just wanting to get all three versions released at the same time so things are synchronized.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -689,5 +689,50 @@ jobs:
...
+      - name: 'Download release artifacts'
+        uses: actions/download-artifact@v2-preview
+      - name: 'Upload macos64 to release'
+        uses: actions/upload-release-asset@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }}
+          asset_path: ${{ github.workspace }}/obs-studio-macOS-release/${{ env.FILE_NAME }}-rel-macOS.dmg
+          asset_name: obs-studio-macOS-${{ env.OBS_GIT_TAG }}.dmg
```
Comment (cg2121) [inline .github/workflows/main.yml:737]: @jp9000 this is where the filename is formatted.

**Comment by cg2121** [comment]
@jp9000 the release is created in draft mode, so other releases can be uploaded and notes added before becoming public, since drafts are private. The filename formatting is handled right before upload.

**Comment by cg2121** [comment]
Closing, until https://github.com/obsproject/obs-studio/pull/5155 is merged, as it rewrites the Github workflow.

**Comment by cg2121** [comment]
Reopened because the cmake rewrite is now merged.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -479,5 +483,39 @@ jobs:
...
+        id: get_version
+        run: echo ::set-output name=VERSION::${GITHUB_REF/refs\/tags\//}
+
+      - name: 'Download macOS x86_64 artifact'
+        uses: actions/download-artifact@v2
+        with:
+          name: obs-macos-${{ steps.get_version.outputs.VERSION }}-x86_64.dmg
+
+      - name: 'Download macOS arm64 artifact'
+        uses: actions/download-artifact@v2
```
Comment (RytoEX) [inline .github/workflows/main.yml:505]: Use v3 actions.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Just a quick review. @PatTheMav ?

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -479,5 +483,39 @@ jobs:
...
+        with:
+          name: obs-macos-${{ steps.get_version.outputs.VERSION }}-arm64.dmg
+
+      - name: 'Create Release'
+        id: create_release
+        uses: softprops/action-gh-release@1e07f4398721186383de40550babbdf2b84acfc5
+        env:
+          IS_RC: ${{ contains(steps.get_version.outputs.VERSION, 'rc') }}
```
Comment (PatTheMav) [inline .github/workflows/main.yml:513]: Move this check into the "Get the version" step above and rename it to something like "Get Metadata" (see also here: https://github.com/obsproject/obs-plugintemplate/blob/077615678bc8513ed34d9e3fd290abe56822615e/.github/workflows/main.yml#L339). 

Set the RC flag as a step output, so you don't have to use the environment variable workaround.

**Comment by PatTheMav** [comment]
I would like to see the checksums being generated and added to the release (see here: https://github.com/obsproject/obs-plugintemplate/blob/077615678bc8513ed34d9e3fd290abe56822615e/.github/workflows/main.yml#L349).

File: .github/workflows/main.yml (hunk 3 above)
Comment (cg2121) [reply .github/workflows/main.yml]: Done

File: .github/workflows/main.yml (hunk 4 above)
Comment (cg2121) [reply .github/workflows/main.yml]: Done

**Comment by PatTheMav** [comment]
Looks good to me.

**Comment by cg2121** [comment]
Needs #6241 to be merged first, so marked this as draft for now.

**Comment by cg2121** [comment]
Updated to include Linux flatpak and deb

**Comment by RytoEX** [comment]
Now that #6241 has been merged, I'll need to see an example of this in action with current git before approving/merging.

**Comment by cg2121** [comment]
After testing on my fork and fixing some issues, everything seems to be working great now.
Screenshot from 2022-08-01 23-33-51

**Comment by cg2121** [comment]
This has been added with https://github.com/obsproject/obs-studio/commit/96a48e86c3a6cd3d4006d04362f761fdfeb87699

**Pull request closed without merging**